File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    // time complexity ->2^(n+m)
    // space complexity -> O(n+m)
    // for memoization, time complexity -> O(n*m) , space complexity -> O(n*m) + O(n+m)
    int f(int i, int j, vector<vector<int>>& grid, vector<vector<int>>& dp){
        
        if(i<0 || j<0)
            return 1e8;
        if(i==0 && j==0)
            return grid[i][j];
        if(dp[i][j] != -1)
            return dp[i][j];
        
        int upperPath = grid[i][j] + f(i-1, j, grid, dp);
        int leftPath = grid[i][j] + f(i, j-1, grid, dp);
        
        return dp[i][j] = min(upperPath, leftPath);
    }
    
    // time complexity -> O(n*m)
    // space complexity -> O(n*m)
    int minPathSum(vector<vector<int>>& grid) {
        
        int n = grid.size();
        int m = grid[0].size();
        
        vector<int> cur(m, 0), prev(m, 0);
        
        // base case
        for(int i=0; i<n; i++){
            
            for(int j=0; j<m; j++){
                
                if(i == 0 and j == 0){
                    cur[j] = grid[i][j];
                    continue;
                }
                
                int upperPath = 1e8, leftPath = 1e8;
                if(i>0) upperPath = grid[i][j] + prev[j];
                if(j>0) leftPath = grid[i][j] + cur[j-1];
                
                cur[j] = min(upperPath, leftPath);
            }
            
            prev = cur;
            
        }
        
        return prev[m-1];
    }
};

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int solve(int i, int j, int[][] grid, int[][] dp){
        
        if(i<0 || j<0)
            return 1000005;
        
        if(i == 0 && j == 0)
            return grid[i][j];
        if(dp[i][j] != -1)
            return dp[i][j];
        
        int up = grid[i][j] + solve(i-1, j, grid, dp);
        int left = grid[i][j] + solve(i, j-1, grid, dp);
        
        return dp[i][j] = Math.min(up, left);
    }
    
    public int minPathSum(int[][] grid) {
        
        int m = grid.length;
        int n = grid[0].length;
        
        int[][] dp = new int[m][n];
        
        for(int i=0; i<m; i++){
            
            for(int j=0; j<n; j++){
                
                if(i == 0 && j == 0){
                    dp[i][j] = grid[i][j];
                    continue;
                }
                
                int up = 1000005, left = 1000005;
                if(i>=1) up = grid[i][j] + dp[i-1][j];
                if(j>=1) left = grid[i][j] + dp[i][j-1];
                
                dp[i][j] = Math.min(up, left);
            }
        }
        
        return dp[m-1][n-1];
    }
}